### PR TITLE
I11 - Add a general Email module

### PR DESCRIPTION
**Pull request by jsbrittain**
Module can be used to send emails at various stages of a workflow. Credentials are provided by local file, and email transmission is triggered by the presence/arrival of a trigger-file which can be user configured.

To test the module:
- Pull down this branch
- Create a 'results/credentials/credentials.json' file (in accordance with the `workflow/Snakefile` docstring); note that this replicates a separate module which would like to a local file during normal execution.
- run `snakemake --use-cores 1 --use-conda`

To test the module in grapevne:
- Obtain and build the latest version of [GRAPEVNE](https://github.com/kraemer-lab/GRAPEVNE)
- Pull down this branch
- redirect grapevne to the `workflows` folder in this repo (use the local copy, since the module had not yet been merged in)
- connect the module to some [Local file](https://github.com/kraemer-lab/vneyard/tree/main/workflows/Utility/sources) modules in order to provide credential and trigger files.

### CODE REVIEW

**Comment by jsbrittain** [comment]
I've updated the module to permit glob-matched email attachments, and to support monitoring of successive email 'triggers' to permit multiple sends (these are useful when monitoring long jobs); there is a test config file with short timeouts that can be run as `snakemake --cores 4 --configfile config/config_test.yaml`.

File: workflows/Utility/modules/Email/workflow/Snakefile [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            # Remove trigger to allow additional calls
+            Path(send_trigger).unlink()
+            # Collate attachments (files can change between triggers due to glob patterns)
+            attachments = []
+            for attachment in params.Attachments:
+                pattern = str(Path(params.AttachmentsFolder) / attachment)
+                glob_expand = glob.glob(pattern)
+                if glob_expand:
+                    attachments.extend(glob_expand)
+                else:
```
Comment (abhidg) [inline workflows/Utility/modules/Email/workflow/Snakefile:160]: If the file exists, wouldn't glob return a truthy value?

**Review by abhidg — APPROVED** [review]
Looks good!

File: workflows/Utility/modules/Email/config/config.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+params:
+  Credentials:
+    Filename: "credentials.json"
+  Server:
+    "SMTP Server": "smtp.gmail.com"
+    "SMTP Port": 587
+  Message:
+    Subject: "[GRAPEVNE] Workflow Alert"
+    Body: "Your workflow triggered an email alert."
+    Recipients: "jsb1982@gmail.com"
```
[the remark this answers is not in the document]
Comment (jsbrittain) [reply workflows/Utility/modules/Email/config/config.yaml:16]: Oh dear! Yes, well spotted!

**Comment by jsbrittain** [comment]
Squashed the last two commits to remove email from the commit history.

**Review by rowannicholls — COMMENTED** [review]
I know this has already been squashed and merged but I have added two relatively minor comments and, also, try as I might I can't get it to work in GRAPEVNE

File: workflows/Utility/modules/Email/config/config_test.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+input_namespace:
+  credentials: "credentials"
+  send: "send"
+  shutdown: "shutdown"
```
[the remark this answers is not in the document]
Comment (jsbrittain) [reply workflows/Utility/modules/Email/config/config_test.yaml:4]: Thanks for looking this over. Yes, snakemake uses files as triggers in many instances. These usually come from the completion of other rules, but for testing are often either created by hand or linked to specific files on the local system. To make this easier and clearer in GRAPEVNE I have added a `Touch` module #18  (that I have tagged you as reviewer on) that will create these files in the editor and allow them to be tested more easily.

File: workflows/Utility/modules/Email/config/config_test.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+params:
+  Credentials:
+    Filename: "credentials.json"
+  Server:
+    "SMTP Server": "smtp.gmail.com"
+    "SMTP Port": 587
+  Message:
+    Subject: "[GRAPEVNE] Workflow Alert"
+    Body: "Your workflow triggered an email alert."
+    Recipients: "example@email.com"
```
[the remark this answers is not in the document]
Comment (jsbrittain) [reply workflows/Utility/modules/Email/config/config_test.yaml:16]: Nice catch, I have fixed this and will submit a new PR with the change.

**Comment by jsbrittain** [comment]
> I know this has already been squashed and merged but I have added two relatively minor comments and, also, try as I might I can't get it to work in GRAPEVNE

@rowannicholls Thanks for taking a look at this - I think you've found an issue that was initially overlooked. Although everything seemed to work on my test-bed, trying it on a clean install did raise an unexpected issue when testing in GRAPEVNE (the `email` standard library module was not being bundled with the packaged python/snakemake environment). I will add this to the new PR.

In the meantime (and just for clarity) to test the module in grapevne:
- Follow the setup instructions in the Email module's docstring to create an 'app password'. You will need the username and password saved into your credentials.json file.
- Use the `LinkLocalFile` module to provide the credentials file to the Email module.
- Connect the same LinkLocalFile module to the 'send' port (you can use it as a trigger by changing the 'send email' trigger filename in the Email module paramers, or use the new Touch module)

Here's a screenshot of my setup:
Screenshot 2024-06-25 at 14 09 13

**Review by rowannicholls — COMMENTED** [review]
Closing this

**Pull request merged**